File: clients/ios/Lextures/Core/LMS/FeedbackLogic.swift
import Foundation

/// Product feedback helpers (FB3) — gating, validation, and payload building.
enum FeedbackLogic {
    static let maxMessageLength = 5000
    static let source = "ios"
    static let categories = ["bug", "idea", "question", "praise", "other"]

    static func feedbackEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffFeedback
    }

    static func messageValid(_ message: String) -> Bool {
        !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    static func trimmedMessageLength(_ message: String) -> Int {
        message.trimmingCharacters(in: .whitespacesAndNewlines).count
    }

    static func appVersion() -> String {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0"
        let build = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "1"
        return "\(version) (\(build))"
    }

    static func buildSubmitRequest(
        message: String,
        category: String,
        route: String,
        locale: String?,
        viewport: String?
    ) -> SubmitFeedbackRequest {
        SubmitFeedbackRequest(
            message: message.trimmingCharacters(in: .whitespacesAndNewlines),
            source: source,
            appVersion: appVersion(),
            context: FeedbackContextPayload(route: route, locale: locale, viewport: viewport),
            category: category.trimmingCharacters(in: .whitespacesAndNewlines)
        )
    }

    enum SubmitOutcome: Equatable {
        case success
        case rateLimited
        case offline
        case error
    }

    static func mapSubmitError(_ error: Error, isOnline: Bool) -> SubmitOutcome {
        if !isOnline {
            return .offline
        }
        if case APIError.httpStatus(429, _) = error {
            return .rateLimited
        }
        if case APIError.transport = error {
            return .offline
        }
        return .error
    }

    static func errorMessageKey(for outcome: SubmitOutcome) -> String {
        switch outcome {
        case .success:
            return "mobile.feedback.success"
        case .rateLimited:
            return "mobile.feedback.rateLimited"
        case .offline:
            return "mobile.feedback.offline"
        case .error:
            return "mobile.feedback.error"
        }
    }

    static func categoryLabelKey(_ category: String) -> String {
        let trimmed = category.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "mobile.feedback.category.none" }
        switch trimmed {
        case "bug", "idea", "question", "praise", "other":
            return "mobile.feedback.category.\(trimmed)"
        default:
            return "mobile.feedback.category.other"
        }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/FeedbackLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.BuildConfig
import com.lextures.android.R
import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.core.network.ApiError

/** Product feedback helpers (FB3). */
object FeedbackLogic {
    const val MAX_MESSAGE_LEN = 5000
    const val SOURCE = "android"
    val CATEGORIES = listOf("bug", "idea", "question", "praise", "other")

    fun feedbackEnabled(features: MobilePlatformFeatures): Boolean = features.ffFeedback

    fun messageValid(message: String): Boolean = message.trim().isNotEmpty()

    fun trimmedMessageLength(message: String): Int = message.trim().length

    fun appVersion(): String = BuildConfig.VERSION_NAME

    fun buildSubmitRequest(
        message: String,
        category: String,
        route: String,
        locale: String?,
        viewport: String?,
    ): SubmitFeedbackRequest {
        val trimmedCategory = category.trim()
        return SubmitFeedbackRequest(
            message = message.trim(),
            source = SOURCE,
            appVersion = appVersion(),
            context = FeedbackContextPayload(
                route = route,
                locale = locale?.takeIf { it.isNotBlank() },
                viewport = viewport?.takeIf { it.isNotBlank() },
            ),
            category = trimmedCategory.takeIf { it.isNotEmpty() },
        )
    }

    sealed class SubmitOutcome {
        data object Success : SubmitOutcome()
        data object RateLimited : SubmitOutcome()
        data object Offline : SubmitOutcome()
        data object Error : SubmitOutcome()
    }

    fun mapSubmitError(error: Throwable, isOnline: Boolean): SubmitOutcome {
        if (!isOnline) return SubmitOutcome.Offline
        if (error is ApiError.HttpStatus && error.code == 429) return SubmitOutcome.RateLimited
        if (error is ApiError.Transport) return SubmitOutcome.Offline
        return SubmitOutcome.Error
    }

    fun errorMessageRes(outcome: SubmitOutcome): Int = when (outcome) {
        SubmitOutcome.Success -> R.string.mobile_feedback_success
        SubmitOutcome.RateLimited -> R.string.mobile_feedback_rateLimited
        SubmitOutcome.Offline -> R.string.mobile_feedback_offline
        SubmitOutcome.Error -> R.string.mobile_feedback_error
    }

    fun categoryLabelRes(category: String): Int = when (category.trim()) {
        "bug" -> R.string.mobile_feedback_category_bug
        "idea" -> R.string.mobile_feedback_category_idea
        "question" -> R.string.mobile_feedback_category_question
        "praise" -> R.string.mobile_feedback_category_praise
        "other" -> R.string.mobile_feedback_category_other
        else -> R.string.mobile_feedback_category_none
    }
}
